Added authentication with google and namespace support for sfMelodyUserFactory::createUser

tquensen opened the pull request:
There was no useful way to get basic user info for google accounts, so a added some changes:

a new "userinfo" namespace in the google class: https://www.googleapis.com/userinfo
the required scope is "https://www.googleapis.com/auth/userinfo.email", so this does not work with the api config (for whatever reason, scope is different to namespace)

As this is not the default namespace, i added  a parameter "namespace" to use with the user configuration.

So, to authenticate with google and retrieve the email, use this config:

```
  melody:
    ...

    google:
      key:  ...
      secret: ...
      callback: "@homepage"   # or absolute url

      # api: []  DOES NOT WORK      # only for google - easy way to set scopes
      scope: [https://www.googleapis.com/auth/userinfo.email]

      user:                       # to create an user
        email:
          call: email
          path: data.email
          key: true
          namespace: userinfo # use the namespace userinfo for the email-call, not the default one
```

Commit:
added userinfo api namespace

Commit:
added namespace option create user calls

Commit:
fixed default namespace

Commit:
Added google plus namespace and removed the 'me' Alias for google (didn't work anyway)

(Namespace requires the scope "https://www.googleapis.com/auth/plus.me" in the app.yml (again, this does not work with the "api" config param as namespace and scope are different)

Access own (public) google+ Profile info via $melody->ns('plus')->get('people/me');

tquensen commented:
Works now with google+ :)

Commit:
Implemented getIdentifier() for google melody.
Uses the google+ ID if available, the email otherwise (or null if none is available)

chok commented on lib/melody/sfGoogleMelody.class.php:
```diff
@@ -37,7 +39,7 @@ protected function initialize($config)
     $this->setCallParameter('alt', 'json');
     $this->setAlias('contacts', 'm8/feeds/contacts');
-    $this->setAlias('me', 'default/full');
+    //$this->setAlias('me', 'default/full');
```
why remove this ?

tquensen replied:
the problem was, that a call to $melody->get('people/me') was changed to 'people/default/full'

chok commented on lib/melody/sfGoogleMelody.class.php:
```diff
@@ -46,6 +48,20 @@ protected function initialize($config)
+  public function getIdentifier()
+  {
+     $prevNs = $this->getCurrentNamespace();  
+     if ($data = $this->ns('plus')->get('people/me')) {
```
this need special authorization ?

but maybe we can check the api params should be better than 2 call to google which can be useless.

What do you think about that ?

tquensen replied:
I have changed that :) see below - should be much better now

chok commented on lib/user/sfMelodyUserFactory.class.php:
```diff
@@ -111,13 +112,14 @@ protected function explodeConfig($config)
       $path = isset($config['path'])?$config['path']:null;
       $prefix = isset($config['prefix'])?$config['prefix']:null;
       $suffix = isset($config['suffix'])?$config['suffix']:null;
+      $ns = isset($config['namespace'])?$config['namespace']:'default';
```
I like that ! very usefull :)

Commit:
Added oauth2 namespace and changed getIdentifier() method

use scope: [https://www.googleapis.com/auth/userinfo.email, https://www.googleapis.com/auth/userinfo.profile] in app.yml and call $melody->ns('oauth2')->get('userinfo') to get the userdata and the unique iserid even if the requested account doesn't have a google+ profile

tquensen commented:
I have changed the getIdentifier method to use only a single call.
The required scope is "https://www.googleapis.com/auth/userinfo.profile" (and optionally "https://www.googleapis.com/auth/userinfo.email" for the email")
I also added a new namespace 'oauth2' => 'https://www.googleapis.com/oauth2/v1' to get the userinfo.
https://www.googleapis.com/oauth2/v1/userinfo doesn't return as much data as a call to the google+ profile, but you get a unique identifier for every google user, even for those without google+ profile

zarpax commented:
I´m using this file version but doesn´t works

tquensen commented:
@zarpax what exactly "doesn´t work"?
Some more information would be helpful :)

zarpax commented:
@tquensen, Sorry for the delay. I've been making new tries and all works ok. In my first try I couldn't connect with google (Token error)
